src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1203,6 +1213,29 @@ internal static int ComputeHashCode(this NamedTypeSymbol type)
                 code++;
             }
             return code;
+
+            static bool wasConstructedForAnnotations(NamedTypeSymbol type)
+            {
+                do
+                {
+                    var typeArguments = type.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics;
+                    var typeParameters = type.ConstructedFrom.TypeParameters;
+
+                    for (int i = 0; i < typeArguments.Length; i++)
+                    {
+                        if (!typeArguments[i].Type.OriginalDefinition.Equals(typeParameters[i], TypeCompareKind.CLRSignatureCompareOptions))
```

Comment:
>typeParameters[i] [](start = 77, length = 17)

Consider calling equality of off that instead of ```typeArguments[i].Type.OriginalDefinition```. I.e. switching sides, this way we know it will always call code for TypeParameters.